### ISSUE TITLE
Only 50% of the main memory can be used for ADIOS2

### ISSUE THREAD

**Issue by psychocoderHPC** [ornladios/ADIOS2#2629]
We saw the following [error message in PIConGPU](https://github.com/ComputationalRadiationPhysics/picongpu/issues/3506#issuecomment-781625774)

```
terminate called after throwing an instance of 'std::_Nested_exception<std::runtime_error>'
  what():  ERROR: buffer overflow when resizing to 36196113975 bytes, in aggregation, when resizing receiving buffer to size 36196113975
```

I was checking the code and think the error is coming from resizing the internal ADIOS buffer  https://github.com/ornladios/ADIOS2/blob/94c2e377eba2947ae9739c1dd42f616a1296f12a/source/adios2/helper/adiosMemory.inl#L549-L550

The numbers in the next example are virtual to illustrate the problem.
I think the problem is that we run on a system with e.g. 48GiB main memory and the internal ADIOS2 buffer had already a size of 28GiB. During writing variables, there was an internal buffer resize to 36GiB required. This failed because to resize the internal ADIOS buffer the `std::vector::resize()` there is not enough main memory (requires 28GiB + 36GiB)  to create the new buffer before the old data is copied over.

Could someone of the ADIOS2 developers please say if I am correct with my example? This means that an internal ADIOS buffer resize is not possible if already more than 50% of the main memory is used by ADIOS+Application if there is no swap partition in the OS available.

I think to remember that we discussed this behavior 3 years ago with @ax3l, @pnorbert, and Scott during the ADIOS2 workshop in Dresden. Unfortunately, I do not remember the outcome therefore I opened this issue to discuss it again.

**Desktop (please complete the following information):**
 - OS/Platform: CentOS Linux release 7.4.1708 (Core)  (Hemera Cluster at HZDR Dresden)
 - ADIOS 2.7.1

**Comment by pnorbert** [comment]
I don't remember either. But your assessment is correct. The vector resize fails. 

Aggregation needs to resize the existing buffer to fit the incoming buffer from another process, which happens to be bigger in this case. There is only one aggregation routine in ADIOS2, so you have no choice but to turn off aggregation by writing one file per process (AggregatorRatio=1 or NumAggregators=100000).

**Comment by psychocoderHPC** [comment]
Thanks for the fast reply.

In general, it should be possible that ADIOS is using an own *vector* implementation that uses native OS function calls and handles physical pages and virtual memory to avoid such an issue if the buffer is resized.
If ADIOS is calling `std::vector::reserve()` with a very large number during the creation of the internal vector this should do the same but it depends strongly on the STL implementation and is not guaarenteed :-(

**Comment by ax3l** [comment]
Yep, related also to #1814

**Comment by williamfgc** [comment]
Any chance that the Put signature returning a span helps in this situation? Essentially, the BP engine would pre-allocate and expose the portion of its buffer as `Variable<T>::Span`. It can be reused in memory-bound applications. See [here](https://adios2.readthedocs.io/en/latest/components/components.html#put-modes-and-memory-contracts).

**Comment by psychocoderHPC** [comment]
> Any chance that the Put signature returning a span helps in this situation? Essentially, the BP engine would pre-allocate and expose the portion of its buffer as `Variable<T>::Span`. It can be reused in memory-bound applications. See [here](https://adios2.readthedocs.io/en/latest/components/components.html#put-modes-and-memory-contracts).

I am not sure how the `Span` can help, could you give more details. 
If I understand it correctly the `Span` will point to an internal buffer of Adios and can be filled by the user instead of passing a flat array with user data into ADIOS. The problem that ADIOS is internally still using a `std::vertor` will stay unchanged. In the case where ADIOS need to resize this internal buffer we have still the problem that we try to fit the old + the new ADIOS Buffer into the main memory.

**Comment by williamfgc** [comment]
> If I understand it correctly the Span will point to an internal buffer of Adios and can be filled by the user instead of passing a flat array with user data into ADIOS

Correct, that way there is no need to duplicate the memory from the application (either from arrays or non-contiguous memory into a contiguous temporary)  into the ADIOS buffer. I don't recall if we tested span with aggregation, though.

> The problem that ADIOS is internally still using a std::vector will stay unchanged. 

`std::vector` is nothing but 3 pointers + functionality around those. The memory strategy won't be very different from using a raw pointer with `new/delete` or `malloc/free`. There is `shrink_to_fit` and there is an option to pass a custom allocator, but that would have to be platform specific and I'm not sure we can do better than the default behavior. 

> In the case where ADIOS need to resize this internal buffer we have still the problem that we try to fit the old + the new ADIOS Buffer into the main memory.

Yes, the BP buffer is built with performance in mind on HPC clusters in which local node memory is "reserved". Adding deallocations comes with a cost as the OS needs to not just wipe old memory, but find contiguous virtual memory for the larger buffer. Just making the (already contiguous) buffer grow is more cost/effective if the application is not memory-bound (that's why `std::vector` `reserve` never deallocates). 

It seems that memory-bound applications have different deallocation requirements, but it won't be cheap to manage memory regardless of the allocator of choice. @pnorbert would it make sense to have an API call that would wipe the buffer either as a hint or mandatory. That should be straight-forward catching the current exception and using `std::vector` `swap`, but costly.  

Hope this helps.
